File: linked_list/reverse_nodes/24_swap_nodes_in_pairs.cc
#include <iostream>

#include "list_node.h"

// Iteration: 引入指针node1和node2, 不容易出错.
// 需要画图之后再进行编程.
ListNode* SwapPairs(ListNode* head) {
  ListNode* dummy_node = new ListNode(-1);
  ListNode* temp = dummy_node;
  temp->next = head;
  while (temp->next && temp->next->next) {
    ListNode* node1 = temp->next;
    ListNode* node2 = temp->next->next;
    temp->next = node2;
    node1->next = node2->next;
    node2->next = node1;
    temp = node1;
  }
  return dummy_node->next;
}

// 2024/01/07: 不看答案写出来了.
// Recursion
ListNode* SwapPairs2(ListNode* head) {
  if (!head || !head->next) {
    return head;
  }

  ListNode* new_head = head->next;
  head->next = SwapPairs2(new_head->next);
  new_head->next = head;
  return new_head;
}

// 2024/01/01自己实现了一个类似但不完全同的递归版本.
ListNode* SwapPairs3(ListNode* head) {
  if (head == nullptr || head->next == nullptr) {
    return head;
  }
  ListNode* prev = head;
  head = head->next;
  prev->next = SwapPairs3(head->next);
  head->next = prev;
  return head;
}

File: linked_list/reverse_nodes/24_swap_nodes_in_pairs.py
from linked_list.list_node import ListNode, create_list_node, print_list_node
from typing import Optional


# 2026/05/14: 自己实现递归.
def swap_pairs(head: Optional[ListNode]) -> Optional[ListNode]:
    if head is None or head.next is None:
        return head

    prev = head
    curr = prev.next
    next = curr.next
    curr.next = prev
    prev.next = swap_pairs(next)
    return curr


head = [1]
head = create_list_node(head)
print_list_node(swap_pairs(head))
